components/suggest/src/rs.rs

```diff
@@ -65,27 +65,96 @@ pub(crate) const DEFAULT_RECORDS_TYPES: [SuggestRecordType; 9] = [
+    ) -> Result<Vec<SuggestRemoteSettingsRecord>> {
+        let options = request.into();
+        self.get_records_with_options(&options)?
+            .records
+            .into_iter()
+            .map(|record| {
+                let attachment_data = record
+                    .attachment
+                    .as_ref()
+                    .map(|a| self.get_attachment(&a.location))
+                    .transpose()?;
+                Ok(SuggestRemoteSettingsRecord::new(record, attachment_data))
+            })
+            .collect()
     }
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Eq, Hash)]
+pub struct SuggestRemoteSettingsRecordRequest {
+    pub record_type: Option<String>,
+    pub last_modified: Option<u64>,
+    pub limit: Option<u64>,
+}
+
+impl From<SuggestRemoteSettingsRecordRequest> for GetItemsOptions {
+    fn from(value: SuggestRemoteSettingsRecordRequest) -> Self {
+        let mut options = GetItemsOptions::new();
 
-    fn get_attachment(&self, location: &str) -> Result<Vec<u8>> {
-        Ok(remote_settings::Client::get_attachment(self, location)?)
+        // Remote Settings returns records in descending modification order
+        // (newest first), but we want them in ascending order (oldest first),
+        // so that we can eventually resume downloading where we left off.
+        options.sort("last_modified", SortOrder::Ascending);
+
+        if let Some(record_type) = value.record_type {
+            options.filter_eq("type", record_type);
+        }
+
+        if let Some(last_modified) = value.last_modified {
+            options.filter_gt("last_modified", last_modified.to_string());
+        }
+
+        if let Some(limit) = value.limit {
+            // Each record's attachment has 200 suggestions, so download enough
+            // records to cover the requested maximum.
+            options.limit((limit.saturating_sub(1) / SUGGESTIONS_PER_ATTACHMENT) + 1);
+        }
+        options
+    }
+}
+
+/// Remote settings record for suggest.
+///
+/// This is `remote_settings::RemoteSettingsRecord`, plus the downloaded attachment data.
+#[derive(Clone, Debug, Default)]
+pub struct SuggestRemoteSettingsRecord {
+    pub id: String,
+    pub last_modified: u64,
+    pub deleted: bool,
+    pub attachment: Option<Attachment>,
+    pub fields: RsJsonObject,
+    pub attachment_data: Option<Vec<u8>>,
+}
+
+impl SuggestRemoteSettingsRecord {
+    pub fn new(record: RemoteSettingsRecord, attachment_data: Option<Vec<u8>>) -> Self {
+        Self {
+            id: record.id,
+            deleted: record.deleted,
+            fields: record.fields,
+            last_modified: record.last_modified,
+            attachment: record.attachment,
+            attachment_data,
+        }
+    }
+
+    pub fn get_attachment_data(&self) -> Result<&[u8]> {
+        self.attachment_data
+            .as_deref()
+            .ok_or_else(|| Error::MissingAttachment(self.id.clone()))
```

Comment:
Good point.  I'm going to rename this `require_attachment_data` since it should only be used in cases where the it's an error for the attachment to be missing.  Maybe we could add a `get_attachment_data` later than returns an Option.